nativelib/src/main/resources/posix/sys/socket.c

```diff
@@ -0,0 +1,123 @@
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <string.h>
+#include <stdio.h>
+
+struct scalanative_sockaddr {
+    char sa_data[14];
+    sa_family_t sa_family;
+};
+
+void scalanative_convert_sockaddr(struct scalanative_sockaddr *in,
+                                  struct sockaddr *out) {
+    out->sa_family = in->sa_family;
+    void *ignored = memcpy(out->sa_data, in->sa_data, 14);
```

Comment:
[The posix spec for `struct sockaddr`](http://pubs.opengroup.org/onlinepubs/000095399/basedefs/sys/socket.h.html#tag_13_61) says that `sockaddr.sa_data` has a variable length. However, [its length has been `14` since the kernel uses git (at least)](https://github.com/torvalds/linux/blob/master/include/linux/socket.h#L29-L32).

[The length also appears to be `14` in Mac OS](https://github.com/opensource-apple/xnu/blob/10.12/bsd/sys/socket.h#L591-L595), though they say that it is actually larger. I'm not sure how to read that comment, but I've been using that code on both MacOS and Linux without any issue.

Indeed once we touch a `sockaddr_storage` or `sockaddr_in6` we'll lose some data. Another solution would be to define, in Scala Native, `sockaddr` as `CStruct0`, have it completely opaque to Scala code, and define functions in C to extract information from that `CStruct0`. I think that's the way to go.

Thanks for pointing out these issues!